### PR TITLE
Add missing generic repeat method

### PR DESCRIPTION
**Pull request by AlexRobson**
Follow up from #460 

Adds in the missing `function rrule(::typeof(repeat), xs::AbstractArray, counts::Integer...)` method. 

Uses the code provided by @mcabbott in the gist [here](https://gist.github.com/mcabbott/80ac43cca3bee8f57809155a5240519f)

Some comments:
- ~There were some type infererence issues in running in 1.0 IDK how important this is is (a bunch of stuff is marked to only check inference in later versions), but to avoid regressing on the tests already implemented I've added some type hinting in the closure. It's a bit gross so there may be a better way to do it.~ EDIT: Now removed and just relaxing the 1.0 tests for typing. 
- ~Bumps finite differences so `rand_tangent{Array{T, 0}}` is now doing the correct thing and hence enables those tests. There are some differences in the behaviour of the forward pass of repeat between 1.0 and 1.6 (the CI versions) so there is a bit of version checking.~ FD has been bumped, but actually no longer needed now that rand_tangent is in CRTU

### CODE REVIEW

**Comment by mcabbott** [comment]
> now doing the correct thing

Meaning what? 

In https://github.com/JuliaDiff/ChainRules.jl/pull/460#issuecomment-871585903 I was trying to say that it seems a bit of an open question what *should* be returned for zero-arrays. They shouldn't error, but should they always be mapped to scalars, or always restored, or is a random mix OK?

**Comment by AlexRobson** [comment]
> > now doing the correct thing
> 
> Meaning what?
> 
> In [#460 (comment)](https://github.com/JuliaDiff/ChainRules.jl/pull/460#issuecomment-871585903) I was trying to say that it seems a bit of an open question what _should_ be returned for zero-arrays. They shouldn't error, but should they always be mapped to scalars, or always restored, or is a random mix OK?

So the issue I was encountering was in TestUtils and the use of `rand_tangent`, which essentially led to the test input to the pullback being a scalar, over a zero-vector.  I don't have the stack trace at hand, but by including these zero-array tests as edge-cases, It was essentially these type conversions which was what was causing these to error. I believe we would either need to promote (somewhere) the zero-vector to a scalar, or as what's been implemented here (or rather in FiniteDifferences), allow the zero-array to propagate. 

I suppose that my stating that it is doing the correct thing is a bit presumptive.  What the 'right' choice of the zero-array idiom is idk. Julia itself seems to have changed the behaviour between versions, i.e. in 1.0 `fill(1.0) + fill(1.0) = 2.0`, and in 1.6 it is `fill(2.0)`. Similar for multiplication - in 1.6 it propagates through (fill(1.0) * 5 == fill(5.0)` (but not if you broadcast 🤷 ). 

So yeah, I'm not certain on what behaviour is wanted. FWIW, I'd expect it to propagate: consistently treat a zero-vector as a vector, and a scalar as a scalar. But idk.

**Comment by mcabbott** [comment]
> FWIW, I'd expect it to propagate: consistently treat a zero-vector as a vector

I'm torn actually. This has the plus side of being easy to state. And sometimes, a zero-array might be a sensible one-number container when you need something mutable and don't want `Ref`? (But why not `Ref`?)

On the minus side, it's a much heavier object than Ref or Float64. And it's a pain to preserve in all rules, since e.g. `A[1,:,1]` is a vector but `A[1,1,1]` is not a zero-array. If you try to restore them from a number, e.g.  like this https://github.com/JuliaDiff/ChainRulesCore.jl/pull/391/commits/d0419b33b0b46205a07f82223929902ccd7676d2, then `fill(1.2)` will always make an `Array` even if you had say a `CuArray`, which will not go well, whereas a number might.

**Comment by AlexRobson** [comment]
Coming back to this MR. 

I've now added in Project for xs and wrapped the calculation of x̄ into a thunk and rebased onto master. 

Responding to your comment (if it's still applies!)

Regarding the consistency - behaviour of zero-arrays, Ref and so on and the nuisance inconsistencies in terms of whether it's treated as a vector or a real this seems a broader question than related to `repeat`. However, if I understand correctly, for where inconsistencies do emerge, Project should allow for some robustness, at least from the perspective of the output of Xrules, as project can just convert the real back into the zero-array (or whatever). So if the pullback did happen to e.g go from a zero-array -> a real number, project should at least make it robust to that. 

However, there may be cases where these inconstencies interfere with the internals of CVC (such as `add!!`) and break.  Zero-arrays do have a habit of going missing as is commented (e.g. in 1.0: `fill(1.0) + fill(1.0) = 2.0`). This does seem to cause an issue here because `add!!` is trying to add a zero-array vector to a real in the `add!!` test. I'm guessing from your comment that you may have seen something like this with CuArrays. If this is the case, this seems to warrant a separate issue.

File: src/rulesets/Base/array.jl [hunk 1]
```diff
@@ -62,54 +62,43 @@ end
...
-function rrule(::typeof(repeat), xs::AbstractVecOrMat, m::Integer, n::Integer)
-    d1, d2 = size(xs, 1), size(xs, 2)
-    function repeat_pullback(ȳ)
-        ȳ′ = reshape(ȳ, d1, m, d2, n)
-        return NoTangent(), reshape(sum(ȳ′; dims=(2,4)), (d1, d2)), NoTangent(), NoTangent()
+        dY = unthunk(ȳ)
+        x̄ = @thunk begin
+                size2ndims = ntuple(d -> isodd(d) ? get(S, 1+d÷2, 1) : get(counts, d÷2, 1), 2*ndims(dY))
+                reduced = sum(reshape(dY, size2ndims); dims = ntuple(d -> 2d, ndims(dY)))
+                project_Xs(reshape(reduced, S))
```
Comment (mzgubic) [inline src/rulesets/Base/array.jl:97]: one level of indentation also here

**Review by mzgubic — APPROVED** [review]
LGTM bar a few minor comments. Are you and @mcabbott happy with the `Fill` outcome?

File: src/rulesets/Base/array.jl [hunk 2]
```diff
@@ -62,54 +62,43 @@ end
...
-        Δ′ = zero(xs)
-        S = size(xs)
-
-        # Loop through each element of Δ, calculate source dimensions, accumulate into Δ′
-        for (dest_idx, val) in pairs(IndexCartesian(), dY)
-            # First, round dest_idx[dim] to nearest gridpoint defined by inner[dim], then
-            # wrap around based on original size S.
-            src_idx = [mod1(div(dest_idx[dim] - 1, inner[dim]) + 1, S[dim]) for dim in 1:length(S)]
-            Δ′[src_idx...] += val
+        x̄ = @thunk begin
```
[the remark this answers is not in the document]
Comment (AlexRobson) [reply src/rulesets/Base/array.jl:71]: > we shouldn't @thunk here because there is only one output

I was considering cases where it's part of the expression, but irrelevant to the gradient calculation. Consider any set-up where repeat is used, but just to set-up constants: 

```
foo(x) = identity(x)
function ChainRules.rrule(::typeof(foo), x)
    function foo_pullback(Ȳ)
        println("pb executed!")
        return (NoTangent(), Ȳ)
    end
    return foo(x), foo_pullback
end
```

```
Zygote.gradient(x -> identity(x), rand())
Zygote.gradient(x -> foo(x), rand()) # pb executed
Zygote.gradient(x -> foo(5) + identity(x), rand()) #pb executed
Zygote.gradient(x -> 5, 1.0)
```

Though doing this, I'm confused why the pb is executing in the third example. It seems that the pb execution is wasteful, given the primal evaluates to a constant. Think I'm misunderstanding something here.

[the remark this answers is not in the document]
Comment (mzgubic) [reply src/rulesets/Base/array.jl]: I don't think Zygote is smart enough to eliminate this kind of "dead code" in the backward pass. I don't think it can tell whether the argument (in this case a literal 5) has any dependence on x, so it just executes all of the pullbacks.

For the third example, I would expect it to run into: https://github.com/FluxML/Zygote.jl/blob/7f2b16987dc4eab236aabd889444cabdcc6f9caf/src/compiler/chainrules.jl#L143
so indeed I also find it weird why the pullback is executing.

**Review by mzgubic — APPROVED** [review]
